src/ripple/app/misc/SHAMapStoreImp.cpp

```diff
@@ -566,19 +573,34 @@ SHAMapStoreImp::clearSql(
     std::string const& minQuery,
     std::string const& deleteQuery)
 {
+    assert(deleteInterval_);
     LedgerIndex min = std::numeric_limits<LedgerIndex>::max();
 
     {
-        auto db = database.checkoutDb();
         boost::optional<std::uint64_t> m;
-        *db << minQuery, soci::into(m);
+        JLOG(journal_.trace())
+            << "Begin: Look up lowest value of: " << minQuery;
+        if (auto db = database.checkoutDb())
+            *db << minQuery, soci::into(m);
+        else
+        {
+            assert(false);
+            return false;
+        }
+        JLOG(journal_.trace()) << "End: Look up lowest value of: " << minQuery;
         if (!m)
             return false;
         min = *m;
     }
 
     if (min > lastRotated || health() != Health::ok)
         return false;
+    if (min == lastRotated)
+    {
+        // Micro-optimization mainly to clarify logs
+        JLOG(journal_.trace()) << "Nothing to delete from " << deleteQuery;
+        return true;
+    }
```

Comment:
Actually, we want to delete up to or equal to lastRotated.
For example, the loop on line 609 should be while (min <= lastRotated)
And if min == lastRotated, then break out of the loop/return after the query is executed. Though there may be something more elegant.

Reply:
Are you sure about that? Every other check runs up to, but doesn't include `lastRotated`, and it's documented that way. https://github.com/ripple/rippled/blob/9173a6a7d3a08be4a7e8345bd763802693a8dcdb/src/ripple/app/misc/SHAMapStoreImp.cpp#L360__

Reply:
To be precise, the code should actually execute the query as it is for all values <=  lastRotated
            DELETE FROM <table> WHERE LedgerSeq < <lastRotated>;

The existing code actually does not call the query when min==lastRotated. It should execute the query. Consider that lastRotated is 1000000. We want to delete everything from 1..999999. But imagine that the function is called initially with min set to 1000000. The delete query will not be called. This is in the existing code. Your opimization made it apparent to me upon close inspection.

Once everything < lastRotated has been deleted, perform no more queries. With the existing code, it would mean checking if min==lastRotated after the delete query, then breaking out of the while loop, and checking for min <= lastRotated in the while condition.

Alternately, something like this:

    while (min <= lastRotated)
    {
        LedgerIndex minRange = std::min(lastRotated, min + deleteBatch);
        {
            auto db =  database.checkoutDb ();
            *db << boost::str (formattedDeleteQuery % minRange);
        }
        if (health())
            return true;
        min = minRange;
        if (min < lastRotated)
            std::this_thread::sleep_for (
                    std::chrono::milliseconds (backOff_));
    }

Reply:
But `min` is initialized from the data in the table.
```
    LedgerIndex min = std::numeric_limits<LedgerIndex>::max();

    {
        boost::optional<std::uint64_t> m;
        JLOG(journal_.trace())
            << "Begin: Look up lowest value of: " << minQuery;
        {
            auto db = database.checkoutDb();
            *db << minQuery, soci::into(m);
        }
        JLOG(journal_.trace()) << "End: Look up lowest value of: " << minQuery;
        if (!m)
            return;
        min = *m;
    }
```
So if `lastRotated` is 1,000,000, and the `minQuery` returns `1,000,000`, that means that there are no records in that particular table from 1..999,999. That's why I added the extra check, and why the loop doesn't need to run. We _know_ that the DELETE query won't delete any rows, because we just discovered that there are no rows to delete.

Reply:
You're correct. The change I suggested doesn't need to be made. Also, keeping your original change is OK.